### DISCUSSION TITLE
How to correctly verify if a certificate is still valid with OCSP?

### DISCUSSION THREAD

**Discussion by ranierimazili** [openssl/openssl#21899] [Q&A]
I would like to understand the ocsp process and how to check if a certificate is still valid using openssl.

### 1 - Testing a valid certificate
I have the following cert that's still valid: [valid-cert.cer.txt](https://github.com/openssl/openssl/files/12475458/valid-cert.cer.txt)
Opening it on windows I could see the intermediary issuer (Soluti) and downloaded it: [soluti.cer.txt](https://github.com/openssl/openssl/files/12475491/soluti.cer.txt)

To discover the OCSP url I've performed the following command:

**Command**
`# openssl x509 -noout -ocsp_uri -in valid-cert.crt.txt`
**Result**
`http://ocsp3.acsoluti.com.br`

Now, to check if the certificate is valid I've tried the command below:

**Command**
`# openssl ocsp -issuer soluti.cer.txt -cert valid-cert.cer.txt -url http://ocsp3.acsoluti.com.br`
**Result**
```
Response Verify Failure
800BEC03457F0000:error:13800065:OCSP routines:ocsp_verify_signer:certificate verify error:../crypto/ocsp/ocsp_vfy.c:64:Verify error: unable to get local issuer certificate
800BEC03457F0000:error:13800065:OCSP routines:ocsp_verify_signer:certificate verify error:../crypto/ocsp/ocsp_vfy.c:64:Verify error: unable to get local issuer certificate
valid-cert.cer.txt: good
        This Update: Aug 30 12:25:13 2023 GMT
        Next Update: Aug 31 12:25:13 2023 GMT
```

I couldn't understand pretty well the response, because it says **Response Verify Failure**, but also says **valid-cert.cer.txt: good**.

What is wrong and what is right with this result?

### 2 - Testing an invalid certificate
I also have a certificate that was revoked:
Cert: [invalid-cert.cer.txt](https://github.com/openssl/openssl/files/12475582/invalid-cert.cer.txt)
Issuer: [serasa.cer.txt](https://github.com/openssl/openssl/files/12475583/serasa.cer.txt)

Using the same approach above to discover the OCSP url, when trying to check if the certificate is still valid, I have the following result:

**Command**
`# openssl ocsp -issuer serasa.cer.txt -cert invalid-cert.cer.txt -url http://ocsp.certificadodigital.com.br/serasasslevv10-3`
**Result**
```
Error querying OCSP responder
807B3C6B167F0000:error:1E800079:HTTP routines:OSSL_HTTP_REQ_CTX_nbio:missing content type:../crypto/http/http_client.c:725:expected=application/ocsp-response
807B3C6B167F0000:error:1E800067:HTTP routines:OSSL_HTTP_REQ_CTX_exchange:error receiving:../crypto/http/http_client.c:874:server=http://ocsp.certificadodigital.com.br:80
```

Where is the problem in this case?

P.S: Renamed all files to txt extention to be able to attach them here

I really appreciate any help... Thanks!!!

**Comment by gingerbeard-man** [comment]
Regarding the first point there are two aspects:
1. You have asked the OCSP responder if the certificate is valid.  Its answer is "yes, that certificate is good".
2. The response is signed.  OpenSSL wants to verify the signature but it does not have access to the necessary certificates.  You have to provide those with the -CApath, -CAfile or -VAfile parameter to prevent the "Verify Failure".

This works for me:
You also need to download the root certificate if you don't already have it in your system's trust store.  Its URL http://acraiz.icpbrasil.gov.br/ICP-Brasilv10.crt appears in the intermediate certificate.  Then create a file containing the PEM (text) version of the intermediate followed by the root.  (The intermediate you provided is in binary DER format so it must be converted.)
E.g. on Linux

```
$ openssl x509 -in soluti.cer.txt -inform DER -out cafile.pem
$ cat ICP-Brasilv10.crt >> cafile.pem
$ openssl ocsp -issuer soluti.cer.txt -cert valid-cert.cer.txt -url http://ocsp3.acsoluti.com.br -CAfile cafile.pem
Response verify OK
valid-cert.cer.txt: good
	This Update: Aug 31 07:48:53 2023 GMT
	Next Update: Sep  1 07:48:53 2023 GMT
```

Disclaimer:  Using -CAfile in this way is probably not best practice with current versions of OpenSSL.  There is a manual page that explains all the different verification options and trusted/untrusted stores.
